pageserver/src/disk_usage_eviction_task.rs

```diff
@@ -0,0 +1,434 @@
+        if cancel.is_cancelled() {
+            return Ok(IterationOutcome::Cancelled);
+        }
+    }
+
+    // carve out the per-tenant reservation and throw everything else into a global list of LRU candidates
+    let mut lru_candidates = Vec::new();
+    for (tenant_id, layers) in &data {
+        let min_resident_size = task_config.tenant_min_resident_size; // XXX allow per-tenant override, including "unlimited"
+        let mut post_evict_resident_size: u64 =
+            layers.iter().map(|(_, layer)| layer.file_size).sum();
+        for (timeline_id, layer) in layers {
+            if cancel.is_cancelled() {
+                return Ok(IterationOutcome::Cancelled);
+            }
+            if post_evict_resident_size <= min_resident_size {
+                continue;
+            }
+            lru_candidates.push((tenant_id, timeline_id, layer));
+            post_evict_resident_size -= layer.file_size;
+        }
+    }
+
+    // do LRU eviction
+    lru_candidates.sort_unstable_by_key(|(_, _, layer)| layer.last_activity_ts);
+    let mut usage_expected = usage_pre.usage.clone();
+    let mut held_tenants = HashMap::new();
+    let mut held_timelines = HashMap::new(); // cache held timelines; XXX limit size?
+    for (tenant_id, timeline_id, layer) in lru_candidates {
+        let tenant = match held_tenants.entry(tenant_id) {
+            Entry::Occupied(e) => e.into_mut(),
+            Entry::Vacant(e) => {
+                let tenant = match tenant::mgr::get_tenant(*tenant_id, false).await {
+                    Ok(tenant) => tenant,
+                    Err(e) => {
+                        warn! {%tenant_id, "failed to get tenant: {:#}", e};
+                        continue;
+                    }
+                };
+                e.insert(tenant)
+            }
+        };
+        let timeline = match held_timelines.entry(timeline_id) {
+            Entry::Occupied(e) => e.into_mut(),
+            Entry::Vacant(e) => {
+                let timeline = match tenant.get_timeline(*timeline_id, false) {
+                    Ok(timeline) => timeline,
+                    Err(e) => {
+                        warn! {%tenant_id, %timeline_id, "failed to get timeline: {:#}", e};
+                        continue;
+                    }
+                };
+                e.insert(timeline)
+            }
+        };
+        // XXX batching? but, if the batch fails due to one layer, we should retry the remaining ones
+        match timeline.evict_layer(&layer.file_name.file_name()).await {
+            Err(e) => {
+                warn! { %tenant_id, %timeline_id, "failed to evict layer: {:#}", e};
+            }
+            Ok(None) | Ok(Some(false)) => {
+                info! { %tenant_id, %timeline_id, "layer does not exist, likely it was deleted" };
+                // XXX should we account for this in usage_expected.avail_bytes ?
+            }
```

Comment:
> In the latter cause, the L1 layer(s) would consume a similar amount of space as the L0 layer. Correct?

We don't delete while compating though that certainly would be an option so you are correct.

> Alternatively, we could do a (rate-limited) statvfs in this case.

Rate-limited because we cannot trust the statvfs or is very expensive to call?

Reply:
> > In the latter cause, the L1 layer(s) would consume a similar amount of space as the L0 layer. Correct?
> 
> We don't delete while compating though that certainly would be an option so you are correct.

I can no longer understand my response here. Related to this, we learned that `layer_removal_cs` is being taken by evictions so we will wait.

> > In the latter cause, the L1 layer(s) would consume a similar amount of space as the L0 layer. Correct?
> 
> We don't delete while compating though that certainly would be an option so you are correct.
> 
> > Alternatively, we could do a (rate-limited) statvfs in this case.
> 
> Rate-limited because we cannot trust the statvfs or is very expensive to call?

Question about rate-limiting still stands, though I remember we were content with how the [02bfc62](https://github.com/neondatabase/neon/pull/3809/commits/02bfc625475908dd939a67f1ac863df8ece22e59) `Result<Option<bool>>` handling was happening for these.